integration/experiment/power_sweep/gen_policy_recommendation.py

```diff
@@ -0,0 +1,350 @@
+    def __init__(self):
+        "Simple constructor."
+        super().__init__(degree=3)
+
+
+def extract_columns(df, region_filter = None):
+    """
+    Extract the columns of interest from the full report collection
+    dataframe. This returns a dataframe indexed by the power limit
+    and columns 'runtime' and 'energy'. region_filter (if provided)
+    is a container that specifies which regions to include (by default,
+    include all of them)."""
+    df_filtered = df
+    if region_filter:
+        df_filtered = df[df['region'].isin(region_filter.split(','))]
+
+    # these are the only columns we need
+    df_cols = df_filtered[[
+        'POWER_PACKAGE_LIMIT_TOTAL',
+        'host',
+        'runtime (sec)',
+        'package-energy (joules)']] \
+    .rename({
+        'POWER_PACKAGE_LIMIT_TOTAL': 'power_limit',
+        'runtime (sec)': 'runtime',
+        'package-energy (joules)': 'energy'
+    },
+    axis = 1) \
+    .groupby(['power_limit', 'host']).sum().reset_index() \
+    [['power_limit', 'runtime', 'energy']] \
+    .set_index('power_limit')
+
+    return df_cols
+
+
+def dump_stats_summary(df, fname):
+    """
+    Write mean runtime and energy and the standard deviation of 
+    runtime and energy for each power limit in CSV format to the
+    file fname."""
+    means = df \
+            .groupby(level=0) \
+            .mean() \
+            .rename(lambda x: x+"_mean", axis=1)
+    stds = df \
+            .groupby(level=0) \
+            .mean() \
+            .rename(lambda x: x+"_stddev", axis=1)
+    pandas.concat([means, stds], axis=1).to_csv(fname)
+
+
+def policy_min_energy(
+        plrange,
+        enmodel,
+        rtmodel = None,
+        pltdp = None,
+        max_degradation = None):
+    """
+    Find the power limit over the range plrange (list-like) that
+    has the minimum predicted energy usage (according to the energy model
+    enmodel, an instance of PowerLimitModel), subject to the constraint
+    that its runtime does not exceed the runtime at power limit pltdp
+    by more than a factor of (1 + max_degradation), if max_degradation is
+    specified. Returns a dictionary with keys power, runtime, and energy,
+    and values the optimal power limit and the predicted runtime and energy
+    at that limit, respectively."""
+
+    if pltdp is None:
+        pltdp = max(plrange)
+
+    en_predictions = enmodel.batch_test(plrange)
+    if max_degradation is None:
+        # we don't need the runtime model in this case
+        best_energy, best_pl = min(zip(en_predictions, plrange))
+        if rtmodel:
+            best_runtime = rtmodel.test(best_pl)
+        else:
+            best_runtime = None
+        return {'power': best_pl,
+                'runtime': best_runtime,
+                'energy': best_energy}
+    else:
+        rt_predictions = rtmodel.batch_test(plrange)
+        rt_at_tdp = rtmodel.test(pltdp)
+        constrained_values = [
+                (energy, pl, runtime)
```

Comment:
I suppose there is no chance that energy will be equal for two evaluations, but if it were, I would order the tuple

(energy, runtime, pl)

Reply:
I like it, fixed.